File: Code/Chapter 13/C#/SqlTVFExternalAccess/ListDirectory.cs
using System;
using System.Data;
using System.Data.Sql;
using System.Data.SqlTypes;
using Microsoft.SqlServer.Server;
using System.IO;
using System.Collections;

namespace SqlTVFExternalAccess
{
   public partial class UserDefinedFunctions
   {
      [Microsoft.SqlServer.Server.SqlFunction(FillRowMethodName = "FillRow", 
         TableDefinition="FileName nvarchar(4000), FileSize nvarchar(4000)")]
      public static IEnumerable GetDircontents(string DirName)
      {
         DirectoryInfo startDir = new DirectoryInfo(DirName);
         // This implements IEnumerable, so we are done :)
         return startDir.GetFiles();
      }

      public static void FillRow(object row, out string FileName, out string FileSize)
      {
         FileInfo oneFile = (FileInfo)row;
         FileName = oneFile.Name;
         FileSize = oneFile.Length.ToString();
      }
   };
}

//Drop Function GetDircontents 
//Drop Assembly SqlTVFExternalAccess


//Create Assembly SqlTVFExternalAccess
//from
//'C:\PlayArea\SqlTVFExternalAccess\bin\Debug\SqlTVFExternalAccess.dll'
//WITH PERMISSION_SET = EXTERNAL_ACCESS
//GO

//Create Function GetDircontents 
//(
//   @DirName NVARCHAR(4000)
//)
//Returns Table 
//(
//   FileDirName NVARCHAR(4000), 
//   FileSize NVARCHAR(4000)
//)
//As
//External Name
//SqlTVFExternalAccess.[SqlTVFExternalAccess.UserDefinedFunctions].GetDircontents 
//Go

//Select * from dbo.GetDircontents ('C:\WINDOWS\Tasks')
//Go

File: Code/Chapter 13/VB.NET/SqlTVFExternalAccess/ListDirectory.vb
Imports System
Imports System.Data
Imports System.Data.Sql
Imports System.Data.SqlTypes
Imports Microsoft.SqlServer.Server
Imports System.Collections
Imports System.Runtime.InteropServices
Imports System.IO

Namespace SqlTVFExternalAccess
   Partial Public Class UserDefinedFunctions
      <Microsoft.SqlServer.Server.SqlFunction(FillRowMethodName:="FillRow", _
         TableDefinition:="FileName nvarchar(4000), FileSize nvarchar(4000)")> _
      Public Shared Function GetDircontents(ByVal DirName As String) As IEnumerable
         Dim startDir As DirectoryInfo = New DirectoryInfo(DirName)
         ' This implements IEnumerable, so we are done :)
         Return startDir.GetFiles()
      End Function

      Public Shared Sub FillRow(ByVal row As Object, _
         <Out()> ByRef FileName As String, <Out()> ByRef FileSize As String)
         ' break the row into its columnar parts. 
         Dim oneFile As FileInfo = CType(row, FileInfo)
         FileName = oneFile.Name
         FileSize = oneFile.Length.ToString()
      End Sub
   End Class
End Namespace